frodo.js

```diff
@@ -2,111 +2,112 @@
  * Extras Plugin
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?